src/main/java/org/apache/sling/installer/core/impl/tasks/BundleTaskCreator.java

```diff
@@ -71,14 +71,17 @@
 
     private BundleBlackList bundleBlacklist;
 
+    private boolean isMultiVersion;
+
     /**
      * @see org.apache.sling.installer.core.impl.InternalService#init(org.osgi.framework.BundleContext, org.apache.sling.installer.api.ResourceChangeListener, RetryHandler)
      */
     @Override
     public void init(final BundleContext bc, final ResourceChangeListener listener, final RetryHandler retryHandler) {
         this.bundleContext = bc;
         this.retryHandler = retryHandler;
-
+        
+        this.isMultiVersion = Boolean.valueOf(bc.getProperty("sling.installer.multiversion")) ? true : false;
```

Comment:
This way this is a global flag. What about a flag per bundle? That way one can decide per bundle whether it is supposed to exist in multiple versions or not. One could add that information to the task attributes and the JCR/File Provider could decide how to populate that attribute (e.g. via Bundle Manifest Header or via some filename pattern).

Reply:
This doesn't match the scenario this is implemented for - this is supposed to be a generic change as otherwise the upgrade can break certain scenarios you cannot recover. The situation with this flag enabled might cause other issues as you indicate but those are recoverable and can be fixed downstream. The problem with the default behavior in cases wehre multiple versions of bundles are active is that the installer destroys the state in an unrecoverable way.  Subsequently we will address downstream cases.

Sidenote - the Dynamic Classloader case is in fact not really related to this case as multiple bundles (different BSNs) cause the same problem as of today.

Reply:
> The problem with the default behavior in cases wehre multiple versions of bundles are active is that the installer destroys the state in an unrecoverable way.

Can you elaborate a bit (maybe on the JIRA), to me it seems that this has been discussed in some private channels, so I don't really understand this remark.

Reply:
I think we can go with a global switch first, if the use case for more fine grained handling comes up it can be built on top of this.
The PR has a lot of formatting changes. Please remove them so we can focus on just the real code changes. Thanks

Reply:
@cziegeler sorry didn't notice that my IDE was too smart - reverted the formatting changes in 2 steps.

Reply:
@kwin there is no need to elaborate on the concrete scenarios faced - but the basic is pretty simple: as soon as you have more than one version of the same bsn in the game (a scenario that the feature launcher and felix support) the statehandling falls appart and you eventually end up with ignored resources where you can't recover the installer state without  starting over from scratch

Reply:
> the statehandling falls appart and you eventually end up with ignored resources where you can't recover the installer state without starting over from scratch

And this PR fixes that? Can we have a test for that?

Reply:
It allows installation & deinstallation of bundles in different versions without impacting other versions.